src/components/Button/Button.vue

```diff
@@ -0,0 +1,424 @@
+			</span>
+			<span v-if="hasText" class="button-vue__text">{{ text }}</span>
+		</span>
+	</button>
+</template>
+<script>
+import Ripple from '../../directives/Ripple'
+
+export default {
+	name: 'Button',
+
+	directives: {
+		ripple: Ripple,
+	},
+
+	props: {
+		/**
+		 * The text of the button.
+		 */
+		text: {
+			type: String,
+			default: '',
+		},
+
+		/**
+		 * Toggles the disabled state of the button on and off.
+		 */
+		disabled: {
+			type: Boolean,
+			default: false,
+		},
+
+		/**
+		 * Specifies the button type
+		 * Accepted values: primary, error, no-background, warning, success. If left empty,
+		 * the default button style will be applyed.
+		 */
+		type: {
+			type: String,
+			validator(value) {
+				return ['primary', 'no-background', 'error', 'warning', 'success', ''].indexOf(value) !== -1
+			},
+			default: '',
+		},
+
+		/**
+		 * Specifies whether the button should span all the available width.
+		 * By default, buttons span the whole width of the container.
+		 */
+		wide: {
+			type: Boolean,
+			default: false,
+		},
+
+		/**
+		 * Always try to provide an aria-label to your button. Make it more
+		 * specific than the button's title by provide some more context. E.g. if
+		 * the title of the button is "send" in the Mail app, the aria label could
+		 * be "Send email".
+		 */
+		ariaLabel: {
+			type: String,
+			default: '',
+		},
+	},
+
+	data() {
+		return {
+			/**
+			 * Keeps track of whether the element's focus status is due to having
+			 * tabbed to it. We use this to display a thick 'focus outline' only
+			 * when the user is navigating with the keyboard.
+			 */
+			tabbed: false,
+		}
+	},
+
+	computed: {
+		hasText() {
+			return this.text
+		},
+
+		hasIcon() {
+			return this.$slots.default !== undefined
+		},
+
+		iconOnly() {
+			return this.hasIcon && !this.hasText
+		},
+
+		textOnly() {
+			return !this.hasIcon && this.hasText
+		},
+
+		iconAndText() {
+			return this.hasIcon && this.hasText
+		},
+
+		// Classes applied to the button element
+		buttonClassObject() {
+			return {
+				// If icon only, some additional css rules are required
+				'button-vue--icon-only': this.iconOnly,
+				'button-vue--text-only': this.textOnly,
+				'button-vue--icon-and-text': this.iconAndText,
+				[`button-vue--vue-${this.type}`]: this.type,
+				'button-vue--wide': this.wide,
+				'button-vue--tabbed': this.tabbed,
+			}
+		},
+	},
+
+	mounted() {
+		/**
+		 * Always fill either the text prop or the ariaLabel one.
+		 */
+		if (!(this.text && this.ariaLabel)) {
+			console.warn('You need to fill either the text or the ariaLabel props in the button component.')
+		}
+		/**
+		 * Check icon slot on first run.
+		 */
+		this.checkIconSlot()
+		/**
+		 * Register observer that checks the icon slot if programmatically changed.
+		 */
+		const observer = new MutationObserver(this.checkIconSlot)
+		observer.observe(this.$el, {
+			childList: true,
+			subtree: true,
+		})
+		this.observer = observer
```

Comment:
How much is this being called per mount/update? I would kind of like a performance analysis on this. We will quickly have a lot of those buttons. If each of them adds even 50ms of runtime, this will adds up pretty quick.

(Irrelevant if we drop those checks as I suggest we do)